File: downloader.py
import asyncio
import re
from pathlib import Path
from typing import Callable, Optional

import httpx

_SEMAPHORE = 4
_HEADERS = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
    "Referer": "https://www.pinterest.com/",
}


def _safe(s: str) -> str:
    return re.sub(r'[\\/*?:"<>|]', "_", s)


def _dest(output_dir: Path, pin_id: str, board_name: str) -> Path:
    return output_dir / f"{pin_id}_{_safe(board_name)}.mp4"


async def _download_one(
    client: httpx.AsyncClient,
    sem: asyncio.Semaphore,
    pin: dict,
    output_dir: Path,
    progress_cb: Optional[Callable],
) -> tuple[str, str]:
    path = _dest(output_dir, pin["pin_id"], pin["board_name"])
    if path.exists():
        if progress_cb:
            progress_cb("skip", pin, path)
        return ("skip", str(path))
    async with sem:
        try:
            async with client.stream("GET", pin["video_url"]) as r:
                r.raise_for_status()
                with open(path, "wb") as f:
                    async for chunk in r.aiter_bytes(65536):
                        f.write(chunk)
            if progress_cb:
                progress_cb("ok", pin, path)
            return ("ok", str(path))
        except Exception as e:
            if progress_cb:
                progress_cb("error", pin, str(e))
            return ("error", str(e))


async def download_all(
    pins: list[dict],
    output_dir: Path,
    progress_cb: Optional[Callable] = None,
) -> tuple[int, int, list[str]]:
    output_dir.mkdir(parents=True, exist_ok=True)
    sem = asyncio.Semaphore(_SEMAPHORE)
    async with httpx.AsyncClient(headers=_HEADERS, follow_redirects=True, timeout=60.0) as client:
        results = await asyncio.gather(
            *[_download_one(client, sem, pin, output_dir, progress_cb) for pin in pins]
        )
    ok = sum(1 for r in results if r[0] == "ok")
    skip = sum(1 for r in results if r[0] == "skip")
    errors = [r[1] for r in results if r[0] == "error"]
    return ok, skip, errors

File: pindl-uxp/src/downloader.js
// Port of downloader.py — concurrent mp4 download into a UXP folder entry.

const { formats } = require("uxp").storage;

const SEMAPHORE = 4;

const HEADERS = {
  "User-Agent":
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
  Referer: "https://www.pinterest.com/",
};

function safe(s) {
  return s.replace(/[\\/*?:"<>|]/g, "_");
}

function destName(pinId, boardName) {
  return `${pinId}_${safe(boardName)}.mp4`;
}

async function fileExists(folder, name) {
  try {
    await folder.getEntry(name);
    return true;
  } catch {
    return false;
  }
}

async function downloadOne(folder, pin, progressCb) {
  const name = destName(pin.pin_id, pin.board_name);

  if (await fileExists(folder, name)) {
    progressCb && progressCb("skip", pin, name);
    return { status: "skip", value: name };
  }

  try {
    const res = await fetch(pin.video_url, { method: "GET", headers: HEADERS });
    if (!res.ok) throw new Error(`HTTP ${res.status}`);
    // UXP filesystem has no streaming writer; buffer then write.
    const buf = await res.arrayBuffer();
    const file = await folder.createFile(name, { overwrite: true });
    await file.write(buf, { format: formats.binary });

    const nativePath = file.nativePath; // absolute path for Premiere import
    progressCb && progressCb("ok", pin, nativePath);
    return { status: "ok", value: nativePath };
  } catch (e) {
    progressCb && progressCb("error", pin, String(e.message || e));
    return { status: "error", value: String(e.message || e) };
  }
}

// Promise pool capped at SEMAPHORE concurrent downloads.
async function downloadAll(pins, folder, progressCb) {
  const results = [];
  let idx = 0;

  async function worker() {
    while (idx < pins.length) {
      const i = idx++;
      results[i] = await downloadOne(folder, pins[i], progressCb);
    }
  }

  const workers = Array.from(
    { length: Math.min(SEMAPHORE, pins.length) },
    worker
  );
  await Promise.all(workers);

  const ok = results.filter((r) => r.status === "ok");
  const skip = results.filter((r) => r.status === "skip");
  const errors = results.filter((r) => r.status === "error").map((r) => r.value);
  const paths = ok.map((r) => r.value);

  return { ok: ok.length, skip: skip.length, errors, paths };
}

module.exports = { downloadAll };
